home-assistant-custom-components.powercalc: init at 1.20.0

CodedNil opened the pull request:
Added home assistant component powercalc https://github.com/bramstroker/homeassistant-powercalc

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dotlambda reviewed (dismissed):
follow github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions

dotlambda commented:
Please squash.

dotlambda reviewed (commented):
We should run their unit tests.

CodedNil commented:
I've squashed the commits. I tried adding unit tests based on other custom components with them but run into an error that I'm not sure how to resolve, sorry.

```
ImportError while loading conftest '/build/source/tests/conftest.py'.
tests/conftest.py:32: in <module>
    from tests.common import mock_area_registry
tests/common.py:3: in <module>
    import custom_components.test.light as test_light_platform
E   ModuleNotFoundError: No module named 'custom_components.test'
```

CodedNil commented:
> Maybe you need `pytest-homeassistant-custom-component`?

I tried with that and many of the other common ones other components used in the tests but still got that error.
Commits are in the correct order.

dotlambda commented:
This is a good start:
```nix
nativeCheckInputs = [
  aioresponses
  pytest-homeassistant-custom-component
  pytestCheckHook
];

preCheck = ''
  patchShebangs --build tests/setup.sh
  tests/setup.sh
'';
```

CodedNil committed:
home-assistant-custom-components.powercalc: init at 1.20.0

mweinelt commented:
```pytb
bramstroker/powercalc> ImportError while loading conftest '/build/source/tests/conftest.py'.
bramstroker/powercalc> tests/conftest.py:22: in <module>
bramstroker/powercalc>     from custom_components.powercalc.const import (
bramstroker/powercalc> custom_components/powercalc/__init__.py:93: in <module>
bramstroker/powercalc>     from .discovery import DiscoveryManager, DiscoveryStatus
bramstroker/powercalc> custom_components/powercalc/discovery.py:37: in <module>
bramstroker/powercalc>     from .helpers import get_or_create_unique_id
bramstroker/powercalc> custom_components/powercalc/helpers.py:23: in <module>
bramstroker/powercalc>     from custom_components.powercalc.power_profile.power_profile import PowerProfile
bramstroker/powercalc> custom_components/powercalc/power_profile/power_profile.py:14: in <module>
bramstroker/powercalc>     from homeassistant.components.camera import DOMAIN as CAMERA_DOMAIN
bramstroker/powercalc> /nix/store/b3ajfxd513xc325xiwy4nx4xhd2br754-homeassistant-2025.12.4/lib/python3.13/site-packages/homeassistant/components/camera/__init__.py:77: in <module>
bramstroker/powercalc>     from .img_util import (
bramstroker/powercalc> /nix/store/b3ajfxd513xc325xiwy4nx4xhd2br754-homeassistant-2025.12.4/lib/python3.13/site-packages/homeassistant/components/camera/img_util.py:8: in <module>
bramstroker/powercalc>     from turbojpeg import TurboJPEG
bramstroker/powercalc> E   ModuleNotFoundError: No module named 'turbojpeg'
```
```patch
diff --git a/pkgs/servers/home-assistant/custom-components/powercalc/package.nix b/pkgs/servers/home-assistant/custom-components/powercalc/package.nix
index 93b3e5db7490..2b96fcadcadc 100644
--- a/pkgs/servers/home-assistant/custom-components/powercalc/package.nix
+++ b/pkgs/servers/home-assistant/custom-components/powercalc/package.nix
@@ -5,6 +5,7 @@
   nix-update-script,
   numpy,
   # Test dependencies
+  home-assistant,
   pytestCheckHook,
   pytest-homeassistant-custom-component,
   pytest-freezegun,
@@ -30,7 +31,8 @@ buildHomeAssistantComponent rec {
     pytestCheckHook
     aioresponses
     pytest-freezegun
-  ];
+  ]
+  ++ home-assistant.getPackages "camera" home-assistant.python.pkgs;
 
   preCheck = ''
     patchShebangs --build tests/setup.sh

mweinelt commented:
And even then the tests don't work against Home Asisstant 2025.12.4.

```pytb
________ ERROR collecting tests/config_flow/test_virtual_power_wled.py _________
ImportError while importing test module '/build/source/tests/config_flow/test_virtual_power_wled.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/3lll9y925zz9393sa59h653xik66srjb-python3-3.13.9/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tests/config_flow/test_virtual_power_wled.py:21: in <module>
    import custom_components.test.sensor as test_sensor_platform
custom_components/test/sensor.py:15: in <module>
    from homeassistant.const import (
E   ImportError: cannot import name 'POWER_VOLT_AMPERE_REACTIVE' from 'homeassistant.const' (/nix/store/b3ajfxd513xc325xiwy4nx4xhd2br754-homeassistant-2025.12.4/lib/python3.13/site-packages/homeassistant/const.py)
_________________ ERROR collecting tests/strategy/test_wled.py _________________
ImportError while importing test module '/build/source/tests/strategy/test_wled.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/3lll9y925zz9393sa59h653xik66srjb-python3-3.13.9/lib/python3.13/importlib/__init__.py:88: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tests/strategy/test_wled.py:29: in <module>
    import custom_components.test.sensor as test_sensor_platform
custom_components/test/sensor.py:15: in <module>
    from homeassistant.const import (
E   ImportError: cannot import name 'POWER_VOLT_AMPERE_REACTIVE' from 'homeassistant.const' (/nix/store/b3ajfxd513xc325xiwy4nx4xhd2br754-homeassistant-2025.12.4/lib/python3.13/site-packages/homeassistant/const.py)
```

Should probably be just `VOLT_AMPERE_REACTIVE`.

CodedNil commented:
Sorry I'm still figuring out the more complicated git commands, how is that? All the tests are running for me after rebasing on master and I'm running this on my server from this branch and all is functioning.
Not sure how to reproduce those test failures you showed, perhaps they aren't there in these latest commits?

CodedNil commented:
> Updated and fixed the required dependencies. Removed nix-update-script, this is the default already.

Oh so sorry I didn't see this until now, is all good?

mweinelt commented:
Yes, all good. Pushed again.
